typeclass-experiments/src/Gen/Generator.hs

```diff
@@ -0,0 +1,68 @@
+module Gen.Generator where
+
+import Control.Monad.Trans.Class
+import Control.Monad.Trans.State
+import Data.Set (Set)
+import Data.Set qualified as S
+import Data.Text (Text)
+import Data.Text.IO qualified as T
+
+import Common.Match
+import Common.SourceTy
+import Common.Types
+
+data GenError
+  = MatchFail Pat Pat
+  | DeriveFail Pat
+  | MultipleMatches Pat [Pat]
+  | MalformedPatList Pat
+  deriving (Show)
+
+type GenResult = Either GenError Text
+
+-- crude code generator
+data Gen = Gen
+  { genPat :: Pat
+  , runGen :: Set Gen -> Pat -> GenResult
+  }
+
+instance Eq Gen where
+  (Gen p _) == (Gen p' _) = p == p'
+
+instance Ord Gen where
+  (Gen p _) <= (Gen p' _) = p <= p'
+
+genFromDefAndPrint :: Set Gen -> TyDef -> IO ()
+genFromDefAndPrint rules def = do
+  let pat = defToPat def
+  case runGenerator rules pat of
+    Left err -> print err
+    Right t -> T.putStrLn t
+
+runGenerator :: Set Gen -> Pat -> Either GenError Text
+runGenerator gens pat = evalStateT (generator pat) gens
+
+generator ::
+  Pat ->
+  StateT (Set Gen) (Either GenError) Text
+generator pat =
+  get >>= \rules ->
+    case matchRules rules of
+      [] -> lift . Left $ DeriveFail pat
+      [g@(Gen p f)] -> do
+        -- The trick is that we make sure every generator always has access to itself.
+        -- This lets us handle infinite applications of TyCons,
+        -- e.g. Maybe (Maybe (Maybe (Maybe IntP)))
+        let gen = Gen p $ \scope px -> f (S.insert g scope) px
+        modify $ S.insert gen
+        newRules <- get
+        lift $ runGen gen newRules pat
+      xs -> lift . Left $ MultipleMatches pat (map genPat xs)
+  where
+    matchRules :: Set Gen -> [Gen]
+    matchRules xs = S.toList $ S.filter (\gx -> matches (genPat gx) pat) xs
```

Comment:
I sense this is the key to this whole PR, but staring at it for 15minutes doesn't reveal what it does. Please formulate this clearly and provide input output examples and elaborate concisely on the purpose.

Reply:
``` haskell
-- some generators to use for the example 
rustInt :: Gen
rustInt = primGen IntP "i64"

rustString :: Gen
rustString = primGen StringP "String"

rustBool :: Gen
rustBool = primGen BoolP "bool"

rustMaybe :: Gen
rustMaybe = Gen (Maybe _x) $ \scope pat -> case pat of
  Maybe t -> do
    t' <- runGenerator scope t
    pure $ "Option<" <> t' <> ">"
  other -> matchFail (Maybe _x) other

rustList :: Gen
rustList = Gen (List _x) $ \scope pat -> case pat of
  List t -> do
    t' <- runGenerator scope t
    pure $ "Vec<" <> t' <> ">"
  other -> matchFail (List _x) other

myGens = S.fromList [rustInt, rustBool, rustString, rustMaybe, rustList]

-- this is just a convenience function so we don't have to type evalStateT everywhere 
runGenerator :: Set Gen -> Pat -> Either GenError Text
runGenerator gens pat = evalStateT (generator pat) gens

-- convenience 
runMyGens :: Pat -> Either GenError Text 
runMyGens = runGenerator myGens 
```
```
-- ghci 
ghci> runMyGens (List IntP)
Right "Vec<i64>"

ghci> runMyGens (List (List IntP))
Right "Vec<Vec<i64>>"

ghci> runMyGens (List (List (Maybe IntP)))
Right "Vec<Vec<Option<i64>>>"

ghci> runMyGens (Maybe (List (List (Maybe IntP))))
Right "Option<Vec<Vec<Option<i64>>>>"
```

I think that + everything else I wrote in all my other responses should make it clear what the purpose is.